server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamMetadata.java

```diff
@@ -40,53 +41,50 @@ public class DataStreamMetadata implements Metadata.Custom {
-        ImmutableOpenMap<String, DataStream> dataStreams = (ImmutableOpenMap<String, DataStream>) args[0];
-        ImmutableOpenMap<String, DataStreamAlias> dataStreamAliases = (ImmutableOpenMap<String, DataStreamAlias>) args[1];
+        Map<String, DataStream> dataStreams = (Map<String, DataStream>) args[0];
+        Map<String, DataStreamAlias> dataStreamAliases = (Map<String, DataStreamAlias>) args[1];
         if (dataStreamAliases == null) {
-            dataStreamAliases = ImmutableOpenMap.of();
+            dataStreamAliases = Map.of();
         }
         return new DataStreamMetadata(dataStreams, dataStreamAliases);
     });
 
     static {
         PARSER.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> {
-            ImmutableOpenMap.Builder<String, DataStream> dataStreams = ImmutableOpenMap.builder();
+            Map<String, DataStream> dataStreams = new HashMap<>();
             while (p.nextToken() != XContentParser.Token.END_OBJECT) {
                 String name = p.currentName();
                 dataStreams.put(name, DataStream.fromXContent(p));
             }
-            return dataStreams.build();
+            return Collections.unmodifiableMap(dataStreams);
         }, DATA_STREAM);
         PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> {
-            ImmutableOpenMap.Builder<String, DataStreamAlias> dataStreams = ImmutableOpenMap.builder();
+            Map<String, DataStreamAlias> dataStreams = new HashMap<>();
             while (p.nextToken() != XContentParser.Token.END_OBJECT) {
                 DataStreamAlias alias = DataStreamAlias.fromXContent(p);
                 dataStreams.put(alias.getName(), alias);
             }
-            return dataStreams.build();
+            return Collections.unmodifiableMap(dataStreams);
         }, DATA_STREAM_ALIASES);
     }
 
-    private final ImmutableOpenMap<String, DataStream> dataStreams;
-    private final ImmutableOpenMap<String, DataStreamAlias> dataStreamAliases;
+    private final Map<String, DataStream> dataStreams;
+    private final Map<String, DataStreamAlias> dataStreamAliases;
 
-    public DataStreamMetadata(
-        ImmutableOpenMap<String, DataStream> dataStreams,
-        ImmutableOpenMap<String, DataStreamAlias> dataStreamAliases
-    ) {
-        this.dataStreams = dataStreams;
-        this.dataStreamAliases = dataStreamAliases;
+    public DataStreamMetadata(Map<String, DataStream> dataStreams, Map<String, DataStreamAlias> dataStreamAliases) {
+        this.dataStreams = Map.copyOf(dataStreams);
+        this.dataStreamAliases = Map.copyOf(dataStreamAliases);
     }
 
     public DataStreamMetadata(StreamInput in) throws IOException {
         this(
-            in.readImmutableOpenMap(StreamInput::readString, DataStream::new),
-            in.readImmutableOpenMap(StreamInput::readString, DataStreamAlias::new)
+            in.readImmutableMap(StreamInput::readString, DataStream::new),
```

Comment:
The serialization format is the same between the two methods, so this doesn't affect compatibility